src/ripple/rpc/impl/ServerHandlerImp.cpp

```diff
@@ -540,185 +553,194 @@ ServerHandlerImp::processRequest (Port const& port,
-    // Determine role/usage so we can charge for invalid requests
-    Json::Value const& method = jsonRPC [jss::method];
-
-    auto role = Role::FORBID;
-    auto required = RPC::roleRequired(method.asString());
-    if (jsonRPC.isMember(jss::params) &&
-        jsonRPC[jss::params].isArray() &&
-        jsonRPC[jss::params].size() > 0 &&
-        jsonRPC[jss::params][Json::UInt(0)].isObject())
-    {
-        role = requestRole(required, port, jsonRPC[jss::params][Json::UInt(0)],
-            remoteIPAddress, user);
-    }
-    else
+    auto size = jsonOrig.isArray() ? jsonOrig.size() : 1;
+    for (unsigned i = 0; i < size; ++i)
     {
-        role = requestRole(required, port, Json::objectValue,
-            remoteIPAddress, user);
-    }
+        Json::Value const& jsonRPC = jsonOrig.isArray() ? jsonOrig[i] : jsonOrig;
+        /* ---------------------------------------------------------------------- */
+        // Determine role/usage so we can charge for invalid requests
+        Json::Value const& method = jsonRPC [jss::method];
 
-    Resource::Consumer usage;
-    if (isUnlimited(role))
-    {
-        usage = m_resourceManager.newUnlimitedEndpoint(
-            remoteIPAddress.to_string());
-    }
-    else
-    {
-        usage = m_resourceManager.newInboundEndpoint(remoteIPAddress);
-        if (usage.disconnect())
+        auto role = Role::FORBID;
+        auto required = RPC::roleRequired(method.asString());
+        if (jsonRPC.isMember(jss::params) &&
+            jsonRPC[jss::params].isArray() &&
+            jsonRPC[jss::params].size() > 0 &&
+            jsonRPC[jss::params][Json::UInt(0)].isObject())
         {
-            HTTPReply(503, "Server is overloaded", output, rpcJ);
-            return;
+            role = requestRole(required, port, jsonRPC[jss::params][Json::UInt(0)],
+                remoteIPAddress, user);
+        }
+        else
+        {
+            role = requestRole(required, port, Json::Value{Json::objectValue},
+                remoteIPAddress, user);
         }
-    }
 
-    if (role == Role::FORBID)
-    {
-        usage.charge(Resource::feeInvalidRPC);
-        HTTPReply (403, "Forbidden", output, rpcJ);
-        return;
-    }
+        Resource::Consumer usage;
+        if (isUnlimited(role))
+        {
+            usage = m_resourceManager.newUnlimitedEndpoint(
+                remoteIPAddress.to_string());
+        }
+        else
+        {
+            usage = m_resourceManager.newInboundEndpoint(remoteIPAddress);
+            if (usage.disconnect())
+            {
+                HTTPReply(503, "Server is overloaded", output, rpcJ);
+                return;
+            }
+        }
 
-    if (! method)
-    {
-        usage.charge(Resource::feeInvalidRPC);
-        HTTPReply (400, "Null method", output, rpcJ);
-        return;
-    }
+        if (role == Role::FORBID)
+        {
+            usage.charge(Resource::feeInvalidRPC);
+            HTTPReply (403, "Forbidden", output, rpcJ);
```

Comment:
I've done some investigating into what bitcoin does.  Here's what I found:

They _do_ allow batch 2.0 requests.  A few top-level checks are made on the request, such as is it a json array, and does it have an overall authorization to be making requests at all.  If any of these top level requests fail, a failing HTTP status code is returned.  Otherwise HTTP 200 (OK) is returned with an array of responses.  Each response will have json error codes (or not) indicating the success/failure of each request.

Non-batch requests are treated the same:  The HTTP status is an error only if there are a few top-level things that go wrong.  Otherwise success is returned even if something low-level went wrong and is reported within the json reply.  Switching to such a system for non-batch requests would be a breaking change for rippled.

This post is meant to be informational only, to be used as a basis for discussion.  I would like to hear opinions on where we should go from here.